vortex-ipc/src/reader.rs

```diff
@@ -123,22 +132,79 @@ pub struct StreamArrayReader<'a, R: Read> {
     messages: &'a mut StreamMessageReader,
     dtype: DType,
     buffers: Vec<Buffer<'a>>,
+    byte_offset: usize,
 }
 
 impl<'a, R: Read> StreamArrayReader<'a, R> {
     pub fn dtype(&self) -> &DType {
         &self.dtype
     }
 
-    pub fn take(&self, indices: &Array<'_>) -> VortexResult<OwnedArray> {
+    pub fn take(&mut self, indices: &Array<'_>) -> VortexResult<OwnedArray> {
         if !indices
             .statistics()
             .compute_as::<bool>(Stat::IsSorted)
             .unwrap_or_default()
         {
             vortex_bail!("Indices must be sorted to take from IPC stream")
         }
-        todo!()
+
+        if self.byte_offset != 0 {
+            vortex_bail!("Stream has already been (at least partially) consumed")
+        }
+
+        let mut chunks = Vec::new();
+
+        let mut offset = self.byte_offset;
+        let idx_len = indices.len();
+        let mut found_so_far = 0;
+
+        // Continue reading batches from the stream until we either run out or find all indices
+        while let Some(batch) = self.next()? {
+            let left =
+                search_sorted::<u64>(indices, offset as u64, SearchSortedSide::Left)?.to_index();
+            let right = search_sorted::<u64>(
+                indices,
+                offset as u64 + batch.len() as u64,
+                SearchSortedSide::Left,
+            )?
+            .to_index();
+
+            // TODO(@jdcasale): replace this with compute scalar_sum when we've added it
+            let indices_for_batch = slice(indices, left, right)?.flatten_primitive()?;
+            let ptype = indices.dtype().try_into()?;
+            let shifted = match_each_unsigned_integer_ptype!(ptype, |$P| {
+                indices_for_batch
+                    .typed_data::<$P>()
+                    .iter()
+                    .map(|&idx| {
+                        idx  as u64 - offset as u64
+                    })
+                    .collect_vec()
+            });
+
+            let shifted_arr = PrimitiveArray::from(shifted).to_array().to_static();
+            let from_current_batch = take(&batch, &shifted_arr)?;
+            found_so_far += from_current_batch.len();
+            chunks.push(from_current_batch);
+            offset += batch.len();
+            if found_so_far >= indices.len() {
+                break;
```

Comment:
You can't really just break out here since now the stream will be left somewhere in the middle. You have to consume the rest of the array and throw it away so that the next call `StreamReader.next()` is correct.